### PR TITLE
Import DataSource Restructuring (1)

### PR DESCRIPTION
**Pull request by adube**
This PR introduces a restructuring of the data source definition within ngeo and gmf to allow the upcoming new "file" data source type.

In summary:

 * we now have `ngeo.datasource` and `gmf.datasource` namespaces (directories)
 * all classes directly related to data sources were moved in there
 * `ngeo.Datasource` has been replaced by:
  * `ngeo.datasource.DataSource` - the abstract class for a data source
  * `ngeo.datasource.OGC` - a data source that uses OGC services to fetch/display its data
 * `gmf.DataSource` has been replaced by `gmf.datasource.OGC`

The 'externs' definitions have been updated as well.

### CODE REVIEW

**Comment by adube** [comment]
@fredj Ready for review.

**Comment by adube** [comment]
@fredj Do you know why Travis is failing?  It seems to happen to me very often and I never really know why.  If you do know why, I'd like to learn how to understand what's wrong to avoid asking that same question too many times.

**Comment by adube** [comment]
Just a reminder: I still don't know what's wrong with Travis.  Any hint would be welcome.

**Comment by sbrunner** [comment]
on the deployement part:
```
ERR! compile Picked up _JAVA_OPTIONS: -Xmx2048m -Xms512m

ERR! compile 

ERR! compile /home/travis/build/camptocamp/ngeo/contribs/gmf/src/directives/layertree.js:682: ERROR - Bad type annotation. Unknown type gmf.datasource.DataSource

ERR! compile  * @param {gmf.datasource.DataSource} ds Data source to filter.

ERR! compile            ^

ERR! compile 

ERR! compile 

ERR! compile 1 error(s), 0 warning(s)

ERR! compile 

ERR! Process exited with non-zero status, see log for more detail: 1 
```
and
```
ERR! compile Picked up _JAVA_OPTIONS: -Xmx2048m -Xms512m

ERR! compile 

ERR! compile /home/travis/build/camptocamp/ngeo/src/datasource/ogc.js:428: ERROR - variable parent is undeclared

ERR! compile     return parent.attributes;

ERR! compile            ^^^^^^

ERR! compile 

ERR! compile 

ERR! compile 1 error(s), 0 warning(s)

ERR! compile 

ERR! Process exited with non-zero status, see log for more detail: 1 
```

**Comment by adube** [comment]
@sbrunner Thanks.

How were you able to see this?  These errors are not reported by Travis.

For me, locally, `make check` still doesn't work (it's been months since I last saw this working).

**Comment by fredj** [comment]
> How were you able to see this? These errors are not reported by Travis.

They are, but they are a bit hidden. To get the error what I usually do is:
 * open travis build
 * open the failing build
 * click "Row log" button
 * scroll to the bottom of the page

**Comment by adube** [comment]
> return parent.attributes;

`parent` refers to the parent class.  This is not a variable.  How come this fails?  How can I avoid the error?

**Comment by fredj** [comment]
@adube #2974 should fix the compile error

**Comment by adube** [comment]
@fredj The current error reported by Travis seems to be caused something that's not yet implemented in the compiler.  Would you please investigate again?

File: src/services/querent.js [hunk 1]
```diff
@@ -241,10 +243,14 @@ ngeo.Querent = class {
...
-          features = dataSource.wfsFormat.readFeatures(response.data);
+        if (dataSource instanceof ngeo.datasource.OGC) {
+          if (wfs) {
+            features = dataSource.wfsFormat.readFeatures(response.data);
+          } else {
+            features = dataSource.wmsFormat.readFeatures(response.data);
+          }
         } else {
-          features = dataSource.wmsFormat.readFeatures(response.data);
+          features = []
```
Comment (fredj) [inline src/services/querent.js]: semicolon

Comment (fredj) [reply src/services/querent.js]: fixed

**Comment by adube** [comment]
@fredj One last error reported by Travis.  Do you know what's wrong?  That doesn't seem related by my patch.

**Comment by fredj** [comment]
See https://travis-ci.org/camptocamp/ngeo/jobs/287557316#L1271
```
ERR! compile /home/travis/build/camptocamp/ngeo/contribs/gmf/src/datasource/datasourceshelper.js:56: ERROR - Bad type annotation. Unknown type gmf.datasource.DataSource
ERR! compile      * @type {Object.<number, gmf.datasource.DataSource>}
ERR! compile                               ^
ERR! compile 
ERR! compile /home/travis/build/camptocamp/ngeo/contribs/gmf/src/datasource/datasourceshelper.js:75: ERROR - Bad type annotation. Unknown type gmf.datasource.DataSource
ERR! compile    * @return {?gmf.datasource.DataSource} Data source.
ERR! compile                ^
ERR! compile 
ERR! compile /home/travis/build/camptocamp/ngeo/contribs/gmf/src/datasource/datasourceshelper.js:79: ERROR - Bad type annotation. Unknown type gmf.datasource.DataSource
ERR! compile     return /** @type {?gmf.datasource.DataSource} */ (
ERR! compile                        ^
ERR! compile 
ERR! compile /home/travis/build/camptocamp/ngeo/contribs/gmf/src/datasource/datasourceshelper.js:138: ERROR - Bad type annotation. Unknown type gmf.datasource.DataSource
ERR! compile  * @typedef {ol.Collection.<gmf.datasource.DataSource>}
ERR! compile                             ^
ERR! compile 
ERR! compile 4 error(s), 0 warning(s)
ERR! compile 
```

**Pull request merged**